### PR TITLE
Add new samplers

### PR DESCRIPTION
**Pull request by abdulfatir**
*Issue #, if available:*

*Description of changes:*
This PR adds two new samplers: `NInstanceSampler` and `ExpectedNumInstanceWithMinSampler`.
`NInstanceSampler`: sample exactly `N` time points per time series.
`ExpectedNumInstanceWithMinSampler`: same as `ExpectedNumInstanceSampler` but ensures at a minimum number of time points are sampled per time series.

Thanks to @lostella for code and suggestions.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

File: src/gluonts/transform/sampler.py [hunk 1]
```diff
@@ -45,6 +45,26 @@ def __call__(self, ts: np.ndarray) -> np.ndarray:
         raise NotImplementedError()
 
 
+class NInstanceSampler(InstanceSampler):
```
Comment (lostella) [inline src/gluonts/transform/sampler.py]: I'm tempted to say that `NumInstanceSampler` would be a better name: just like `ExpectedNumInstanceSampler` samples a certain number of instances in expectations, this samples a certain number of instances (full stop).

Let's see also what others think about the naming maybe

Comment (lostella) [reply src/gluonts/transform/sampler.py]: Let's say: if all these samplers had a more "assertive" naming, I would just call this `SampleN` (and then `SampleOnAverageN`)

File: src/gluonts/transform/sampler.py [hunk 2]
```diff
@@ -142,6 +162,37 @@ def __call__(self, ts: np.ndarray) -> np.ndarray:
         return indices + a
 
 
+class ExpectedNumInstanceWithMinSampler(ExpectedNumInstanceSampler):
```
Comment (lostella) [inline src/gluonts/transform/sampler.py]: I see two ways to escape the naming dilemma here:
1. Make `min_instances: int = 0` (note the default `0`) an attribute of `ExpectedNumInstanceSampler` and add the logic directly there. Then there's no name to be invented. However, strictly speaking when one sets `min_instances > 0` then this is not really sampling `num_instances` on average.
2. Make this logic a wrapper around some other sampler: `sampler = AtLeast(1, ExpectedNumInstanceSampler)`. The logic is factored a bit better maybe that way, just not as straightforward to instantiate.

Again, no strong opinions, it's just that this names is now a mouthful. @jaheba to the rescue.

File: src/gluonts/transform/sampler.py (hunk 1 above)
Comment (abdulfatir) [reply src/gluonts/transform/sampler.py]: Makes sense. Updated.

File: src/gluonts/transform/__init__.py [hunk 3]
```diff
@@ -33,6 +33,8 @@
     "ContinuousTimePredictionSampler",
     "ExpandDimArray",
     "ExpectedNumInstanceSampler",
+    "ExpectedNumInstanceWithMinSampler",
+    "NInstanceSampler",
```
Comment (lostella) [inline src/gluonts/transform/__init__.py]: These also need updating

File: src/gluonts/transform/sampler.py (hunk 2 above)
Comment (lostella) [reply src/gluonts/transform/sampler.py]: @abdulfatir let's go for option 1

File: src/gluonts/transform/sampler.py [hunk 4]
```diff
@@ -142,6 +162,37 @@ def __call__(self, ts: np.ndarray) -> np.ndarray:
...
+
+    def __call__(self, ts: np.ndarray) -> np.ndarray:
+        indices = super().__call__(ts)
+
+        a, b = self._get_bounds(ts)
+
+        if a > b:
+            return indices
+        if len(indices) < self.min_instances:
+            prefix = np.random.choice(
```
Comment (jaheba) [inline src/gluonts/transform/sampler.py]: `np.random.randint` or `.random_integer` should make this a bit more concise.

File: src/gluonts/transform/sampler.py [hunk 5]
```diff
@@ -142,6 +162,37 @@ def __call__(self, ts: np.ndarray) -> np.ndarray:
...
+
+    num_instances: float
+    min_instances: int = 1
+
+    def __call__(self, ts: np.ndarray) -> np.ndarray:
+        indices = super().__call__(ts)
+
+        a, b = self._get_bounds(ts)
+
+        if a > b:
```
Comment (jaheba) [inline src/gluonts/transform/sampler.py]: Should this be `>=`?

And also, can we return `np.array([])` in that case, I think it makes it a bit more clear what this does then.

File: src/gluonts/transform/sampler.py [hunk 6]
```diff
@@ -142,6 +162,37 @@ def __call__(self, ts: np.ndarray) -> np.ndarray:
...
+            return indices
+        if len(indices) < self.min_instances:
+            prefix = np.random.choice(
+                range(a, b + 1), size=self.min_instances - len(indices)
+            )
+            indices = np.concatenate([prefix, indices])
```
Comment (jaheba) [inline src/gluonts/transform/sampler.py]:
```suggestion
            return np.concatenate([prefix, indices])
```

**Comment by jaheba** [comment]
I think we can just extend the current class with `min_instances=0`

**Review by TonyHCIR — APPROVED** [review]

File: src/gluonts/transform/sampler.py (hunk 2 above)
Comment (abdulfatir) [reply src/gluonts/transform/sampler.py]: Done.

File: src/gluonts/transform/sampler.py (hunk 4 above)
Comment (abdulfatir) [reply src/gluonts/transform/sampler.py]: Done

**Comment by abdulfatir** [comment]
@lostella @jaheba Removed the new class and added `min_instances` option to the original class.

**Review by lostella — APPROVED** [review]

File: src/gluonts/transform/sampler.py [hunk 7]
```diff
@@ -45,6 +45,26 @@ def __call__(self, ts: np.ndarray) -> np.ndarray:
         raise NotImplementedError()
 
 
+class NumInstanceSampler(InstanceSampler):
+    """
+    Samples N time points from each series.
+
+    Parameters
+    ----------
+    N
```
Comment (jaheba) [inline src/gluonts/transform/sampler.py:54]: Why is this uppercase?

**Pull request merged**